File: prob1.java
// Time Complexity : O(nm^2) 
// Space Complexity : O(nm^2)
// Did this code successfully run on Leetcode : Yes

class Solution {
    public List<List<String>> wordSquares(String[] words) {
       
        TrieNode root = new TrieNode();
        for (String word : words) {
             TrieNode node = root;
            for (char ch : word.toCharArray()) {
                node.children.computeIfAbsent(ch, k -> new TrieNode());
                node =  node.children.get(ch);
                node.setWord(word);
            }
        }
        int n = words[0].length();
        List<List<String>> ans = new ArrayList<>();
        for (String word : words) {
            List<String> wordSquares = new ArrayList<String>();
            wordSquares.add(word);
            match(root,wordSquares, ans, 1);
        }
        return ans;
        
    }
    
    private void  match(TrieNode root,  List<String> wordSquares, List<List<String>> ans, int step) {
        int n = wordSquares.get(0).length();
        if (step == n) {
            ans.add(new ArrayList<>(wordSquares));
            return;
        }
        StringBuilder prefix = new StringBuilder();
        prefix = getPrefix(wordSquares, step);
        for (String matchedWord : getWordsWithPrefix(root, prefix.toString())) {
            wordSquares.add(matchedWord);
            match(root, wordSquares, ans, step + 1);
            wordSquares.remove(wordSquares.size() - 1);
        }
    }
    private  List<String> getWordsWithPrefix(TrieNode root, String prefix) {
        TrieNode node = root;
        for (Character letter : prefix.toCharArray()) {
        if (node.children.containsKey(letter)) {
            node = node.children.get(letter);
        } else {
            // return an empty list.
            return new ArrayList<String>();
        }
        }
        return node.words;
  }

    private StringBuilder getPrefix(List<String> wordSquares, int step) {
        StringBuilder prefix = new StringBuilder();
        for (String word : wordSquares) {
            prefix.append(word.charAt(step));
        }
        return prefix;
    }
    
    class TrieNode {
        public Map<Character, TrieNode> children;
        public List<String> words;
        
        public TrieNode() {
            children = new HashMap<>();
        }
        public void setWord(String word) {
            if (words == null) {
                words = new ArrayList<>();
            }
            words.add(word);
        }
    }
}

File: prob1.py
# Time Complexity : O(N * M^2)
# Space Complexity : O(N * M^2)
# Passed on Leetcode: yes

class Solution:
    def wordSquares(self, words):
        def is_prefix(prefix, word):
            return word.startswith(prefix)
        
        def backtrack(square):
            if len(square) == len(words[0]):
                squares.append(list(square))
                return
            
            col = len(square)
            prefix = ''.join(row[col] for row in square)
            
            for word in prefix_to_words[prefix]:
                if is_prefix(''.join(word[col + 1] for col in range(col)), word):
                    square.append(word)
                    backtrack(square)
                    square.pop()
        
        if not words:
            return []
        
        squares = []
        prefix_to_words = {}
        
        # Preprocess the words and build a dictionary of prefixes to words
        for word in words:
            for i in range(len(word)):
                prefix = word[:i]
                if prefix not in prefix_to_words:
                    prefix_to_words[prefix] = []
                prefix_to_words[prefix].append(word)
        
        # Start backtracking
        for word in words:
            square = [word]
            backtrack(square)
        
        return squares